### DISCUSSION TITLE
InstanceOperation with FhirString parameter

### DISCUSSION THREAD

**Discussion by elvetian** [FirelyTeam/firely-net-sdk#1769] [Q&A] [answered]
I trying to send along a `FhirString` parameter with [InstanceOperation](https://github.com/FirelyTeam/firely-net-sdk/blob/c10e1b13521b5e625583bd48643d9f68a05fac79/src/Hl7.Fhir.Core/Rest/BaseFhirClient.cs#L788) but the server side does not read it correctly. 

In [FhirClientTests:759](https://github.com/FirelyTeam/firely-net-sdk/blob/f71543edc34c9edecf0f13af50d35e9e57ca353a/src/Hl7.Fhir.Core.Tests/Rest/FhirClientTests.cs#L759) there is a similar example to mine but the FhirString parameter is commented out:
            `Parameters p = new Parameters();`
            `//  p.Add("mode", new FhirString("create"));`
            `p.Add("resource", pat);`
            `OperationOutcome ooI = (OperationOutcome)client.InstanceOperation(ri.WithoutVersion(), "validate", p);`

Isn't it supported? What is the correct way to do this?

**Comment by marcovisserFurore** [comment, marked as answer]
Hi @elvetian,

Indeed, this is the way to do it. In the `FhirClientTest` we are executing a `validate` operation. That operation demands a code as parameter (see also the definition [here](https://www.hl7.org/fhir/resource-operation-validate.html)). So actually the`FhirClient` request should be like this:
```c#
      Parameters p = new Parameters();
      p.Add("mode", new Code("create"));
      p.Add("resource", pat);
      OperationOutcome ooI = (OperationOutcome)client.InstanceOperation(ri.WithoutVersion(), "validate", p);
```

But, when an operation demands a parameter of type string, then you use the `FhirString` class. See here an example of the operation `$graphql`, ([definition](https://www.hl7.org/fhir/resource-operation-graphql.html)): 
```c#
      Parameters p = new Parameters();
      p.Add("query", new FhirString("{name{text,given,family}}"));
      OperationOutcome ooI = (OperationOutcome)client.InstanceOperation(ri.WithoutVersion(), "graphql", p);
```
It can be that the FHIR server does not accept this operation. Check the OperationOutcome for more details. And try also the same request in Postman for example. It should give you back the same result.

**Reply by elvetian** [reply]
Hi @marcovisserFurore, 

I have this request 
```
http://{{FHIR_SERVER_IP}}:44376/fhir/Appointment/Appointment-8/$checkin?_format=json
```
with the following body 
``` json
{"resourceType":"Parameters", "parameter":[{"name":"locationKey","valueString":"Location-1011"}]}
```
This works well in Postman.


Here's my implemenation with firely-sdk:
``` c#
var parameters = new Parameters();
parameters.Add("resourceType", new FhirString("Parameters"));
parameters.Add("parameter", new FhirString($"[{{\"name\":\"locationKey\",\"valueString\":\"Location-1030\"}}]"));
var appointmentResource = ResourceIdentity.Build("Appointment", "Appointment-8");
client.InstanceOperation(appointmentResource, "checkin", parameters, false);
```
The response I get is
``Operation was unsuccessful because of a client error (UnprocessableEntity). OperationOutcome: Overall result: FAILURE (1 errors and 0 warnings)``
``[ERROR] Resource id not present.``

It seems like the FHIR server is not able to parse the JSON string parameter to get the resource id "Location-1030"?

Thanks for your help!

**Reply by marcovisserFurore** [reply]
Hi @elvetian,

Almost right. There is only 1 parameter here: `locationKey`. So the code would be:

```c#
            var parameters = new Parameters();
            parameters.Add("locationKey", new FhirString("Location-1011"));
            var appointmentResource = ResourceIdentity.Build("Appointment", "Appointment-8");
            client.InstanceOperation(appointmentResource, "checkin", parameters, false);
```

the variable `parameters` is going to be serialized in the `FhirClient` as:

```json
{"resourceType":"Parameters", "parameter":[{"name":"locationKey","valueString":"Location-1011"}]}
```

**Reply by mbaltus** [reply]
@elvetian You do not need to recreate the json format of the Parameters resource. The FhirClient will serialize the Parameters correctly when sending out the request. So this should be enough to setup the Parameters:
```
var parameters = new Parameters();
parameters.Add("locationKey", new FhirString("Location-1030"));
```

If you then still get the same error from the server, you could try and add an id to the Parameters, since that is what the error says is missing:
```
parameters.Id = Guid.NewGuid().ToString();
```

**Reply by elvetian** [reply]
Many thanks @marcovisserFurore and @mbaltus!
This works for me:
``` C#
var parameters = new Parameters();
parameters.Add("locationKey", new FhirString("Location-1030"));
```
